core/trino-main/src/main/java/io/trino/operator/aggregation/state/Int128StateSerializer.java

```diff
@@ -13,58 +13,57 @@
  */
 package io.trino.operator.aggregation.state;
 
-import io.airlift.slice.Slice;
-import io.airlift.slice.Slices;
 import io.trino.spi.block.Block;
 import io.trino.spi.block.BlockBuilder;
 import io.trino.spi.function.AccumulatorStateSerializer;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.Decimals;
 import io.trino.spi.type.Int128;
 import io.trino.spi.type.Type;
 
-import static io.trino.spi.type.VarbinaryType.VARBINARY;
 
-public class LongDecimalWithOverflowStateSerializer
-        implements AccumulatorStateSerializer<LongDecimalWithOverflowState>
+public class Int128StateSerializer
+        implements AccumulatorStateSerializer<Int128State>
 {
-    private static final int SERIALIZED_SIZE = Long.BYTES + Int128.SIZE;
+    private static final int SERIALIZED_SIZE = Long.BYTES + Int128.SIZE * 2;
+    private final DecimalType type;
+
+    public Int128StateSerializer(DecimalType elementType)
+    {
+        this.type = DecimalType.createDecimalType(Decimals.MAX_SHORT_PRECISION + 1);
+    }
 
     @Override
     public Type getSerializedType()
     {
-        return VARBINARY;
+        return type;
     }
 
     @Override
-    public void serialize(LongDecimalWithOverflowState state, BlockBuilder out)
+    public void serialize(Int128State state, BlockBuilder out)
     {
         if (state.isNotNull()) {
-            long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(overflow, decimal[offset], decimal[offset + 1]));
+            type.writeObject(out, Int128.valueOf(decimal[offset], decimal[offset + 1]));
```

Comment:
this can allocate `Int128` for every row if not stack allocated by escape analysis. I would use `out builder` directly.